### DISCUSSION TITLE
EEPROM Bytes overwrite possible?

### DISCUSSION THREAD

**Discussion by tsaG1337** [earlephilhower/arduino-pico#264] [General]
Hi all,

I do use the EEPROM Emulation to store Data on my RP2040. At the beginning, it is checked if an EEPROM State Flag is correct (basically a version number of the EEPROM content). This Flag hardware coded before Firmware Upload. It basically says If Flag != Flag in EEPROM -> Flush EEPROM.

However, it seems that I can't write to the EEPROM at all, the function is called over and over again.

```
const uint8_t ADRESS_START = 0; // start address of the EEPROM
#define EEPROM_SIZE 512
#define EEPROM_STATE_ADRESS EEPROM_SIZE //The State Change Byte is stored in the last field in the EEPROM
const uint8_t EEPROM_STATE = 1; //Change the value to Purge the EEPROM

void setup() {
  Serial.begin(19200);
  EEPROM.begin(EEPROM_SIZE); //We use EEPROM_STATE_ADDRESS as the maximum size of the ROM
}

void loop() {
  
  loadConfig();
  delay(3000);

}

int loadConfig() {

  if (EEPROM.read(EEPROM_STATE_ADRESS) != EEPROM_STATE) {
    DEBUG_PRINTLN("New EEPROM State. Purging old Data");
    // write a 0 to all 512 bytes of the EEPROM
    for (int i = 0; i < EEPROM.length(); i++) {
      EEPROM.write(i, 0);
    }
    EEPROM.write(EEPROM_STATE_ADRESS, EEPROM_STATE); //Write new State to EEPROM
    if (EEPROM.commit()) {
      DEBUG_PRINTLN("EEPROM successfully committed");
      DEBUG_PRINT("EEPROM STATE is: ");
      DEBUG_PRINTLN(EEPROM.read(EEPROM_STATE_ADRESS));
    } else {
      DEBUG_PRINTLN("ERROR! EEPROM commit failed");
    }
}

```
The OUTPUT however is the following

> 13:58:39.298 -> New EEPROM State. Purging old Data

> 13:58:39.334 -> EEPROM successfully committed
> 13:58:39.334 -> EEPROM STATE is: 0
> 13:58:39.334 -> EEProm Config differs from Default.
> 13:58:42.337 -> New EEPROM State. Purging old Data
> 13:58:42.371 -> EEPROM successfully committed
> 13:58:42.371 -> EEPROM STATE is: 0
> 13:58:45.367 -> New EEPROM State. Purging old Data
> 13:58:45.404 -> EEPROM successfully committed
> 13:58:45.404 -> EEPROM STATE is: 0
> 13:58:48.400 -> New EEPROM State. Purging old Data
> 13:58:48.437 -> EEPROM successfully committed
> v13:58:48.437 -> EEPROM STATE is: 0

**Comment by earlephilhower** [comment]
I believe you have an off-by-one error in the code.  You created a fake EEPROM w/512 bytes from 0...511, but are then trying to read/write address 512 which is out of bounds.  Try making `#define EEPROM_STATE_ADDRESS (EEPROM_SIZE-1)`
